source/components/tooltip/HxTooltip.js

```diff
@@ -0,0 +1,191 @@
+import { getPosition } from '../../lib/position';
+import debounce from 'lodash/debounce';
+
+window.addEventListener('WebComponentsReady', function () {
+    const tagName = 'hx-tooltip';
+    const template = document.createElement('template');
+
+    template.innerHTML = `
+      <style>${require('./HxTooltip.less')}</style>
+      ${require('./HxTooltip.html')}`;
+
+    class HxTooltip extends HTMLElement {
+        static get is () {
+            return tagName;
+        }
+
+        static get observedAttributes () {
+            return [ 'open' ];
+        }
+
+        constructor () {
+            super();
+            this.attachShadow({ mode: 'open' });
+            if (window.ShadyCSS) {
+                ShadyCSS.prepareTemplate(template, tagName);
+                ShadyCSS.styleElement(this);
+            }
+            this.shadowRoot.appendChild(template.content.cloneNode(true));
+            this.initialPosition = this.position;
+            this._show = this._show.bind(this);
+            this._hide = this._hide.bind(this);
+            this._toggle = this._toggle.bind(this);
+            this._reposition = this._reposition.bind(this);
+            this._closeOnBackgroundClick = this._closeOnBackgroundClick.bind(this);
+        }
+
+        connectedCallback () {
+            this._target = document.querySelector('[data-tooltip=' + this.id + ']');
+            if (!this.hasAttribute('role')) {
+                this.setAttribute('role', 'tooltip');
+            }
+            if (!this._target) {
+                return;
+            }
+            this._connectHandlers();
+        }
+
+        disconnectedCallback () {
+            if (!this._target) {
+                return;
+            }
+            this._destoryAllHandlers();
+        }
+
+        attributeChangedCallback (attr, oldValue, newValue) {
+            switch (attr) {
+                case 'open':
+                    this.setAttribute('aria-hidden', newValue !== '');
+                    break;
+            }
+        }
+
+        _hide () {
+            if (this._showTimer) {
+                clearTimeout(this._showTimer);
+            }
+            this._hideTimer = setTimeout(() => {
+                this.open = false;
+            }, 1600);
+            this.position = this.initialPosition;
+        }
+
+        _show () {
+            if (this._hideTimer) {
+                clearTimeout(this._hideTimer);
+            }
+            this._showTimer = setTimeout(() => {
+                this.open = true;
+                this._setPosition();
+            }, 500);
+        }
+
+        _toggle () {
```

Comment:
we can probably refactor this to apply `_setPosition()` and `this.position = ...` in the `set position (value)` property to simplify.